### PR TITLE
RDMA builtin support

### PR DESCRIPTION
**Pull request by pizhenwei**
Hi,

There are several patches in this PR:
* Abstract set/rewrite config bind option: `bind` option is a special config, `socket` and `tls` are using the same one. However RDMA uses the similar style but different one. Use a bit abstract work to make it flexible for both `socket` and `RDMA`. Even for QUIC in the future.
* Introduce closeListener for connection type: closing socket by a simple syscall would be fine, RDMA has complex logic. Introduce connection type specific close listener method.
* RDMA: Use valkey.conf style instead of module parameters: use `--rdma-bind` and `--rdma-port` style instead of module parameters. The module style config `rdma.bind` and `rdma.port` are removed.
* RDMA: Support builtin: support `make BUILD_RDMA=yes`. module style is still kept for now. Once module style is decided to drop, I volunteer to do it for TLS and RDMA together.

Each patch has independent functionality, also been tested by CI and local commands, so request no-squash merge for this PR.

### CODE REVIEW

**Comment by zuiderkwast** [comment]
@valkey-io/core-team Please approve.

This adds `make BUILD_RDMA=yes`, so just like TLS it can be built as a module or as builtin. Just like for TLS, the default is `no`.

It is a step towards official (non-experimantal) RDMA support.

I already merged the RDMA fork-safety here: #1244. It makes sure the server doesn't crash if we try to talk to an RDMA client from a forked child process.

**Comment by pizhenwei** [comment]
> @valkey-io/core-team Please approve.
> 
> This adds `make BUILD_RDMA=yes`, so just like TLS it can be built as a module or as builtin. Just like for TLS, the default is `no`.
> 
> It is a step towards official (non-experimantal) RDMA support.
> 
> I already merged the RDMA fork-safety here: #1244. It makes sure the server doesn't crash if we try to talk to an RDMA client from a forked child process.

Hi @valkey-io/core-team
The [CMake building support](https://github.com/valkey-io/valkey/pull/1196) has been merged, then we can drop module build option for TLS and RDMA together.

**Comment by PingXie** [comment]
> The CMake building support has been merged, then we can drop module build option for TLS and RDMA together.

I vote for dropping the module packaging. Modularity is what we need and we have it here for RDMA.

**Comment by zuiderkwast** [comment]
I believe it's safe to drop the RDMA module because it's experimental, but I'm not sure about the TLS module. Maybe someone is using it? Maybe we should drop it in a major version (9.0?).

The potential benefit of module is that you don't need OpenSSL installed to start Valkey, as long as you don't load the module. If we ship binary releases or containers with bundled modules, they can be used also by those who don't have or want TLS. Examples are embedded system, IoT devices, etc. We've already talked about packaging binaries with buldled modules like (bloom, json).

File: valkey.conf
Comment (zuiderkwast) [inline valkey.conf]: Mention the word "default" and use the default value in the example.

```suggestion
# The RDMA receive transfer buffer is 1M by default. It can be set between 64K and 16M.
#
# rdma-rx-size 1048576
```

File: valkey.conf
Comment (zuiderkwast) [inline valkey.conf]: "it must be at least zero" and "The default vector is -1" are contradicting. Very confusing. :smile: Can you fix the text so it's not contradicting itself?

When/why would you want to use a random/non-random vector?

File: valkey.conf
Comment (zuiderkwast) [inline valkey.conf]: It's not important how RDMA is built. It's only important that it *is* built.

But I don't think we need to mention it. TLS build is optional too and we don't mention that for the TLS configs.

```suggestion
# directive can be used to define RDMA-listening ports. To enable RDMA on the
# port, use:
```

**Review by zuiderkwast — COMMENTED** [review]
I didn't review fully. Just the interface changes.

File: valkey.conf
Comment (pizhenwei) [reply valkey.conf]: Sure, add more information about system interrupt pinning, RDMA completion vector and valkey-server cpu affinity to tune performance. Please take a look at the latest version. Thanks!

**Comment by pizhenwei** [comment]
Hi @zuiderkwast 
Please take a loot at the latest version, modify as your suggestion, and support builtin for cmake.

File: valkey.conf
Comment (zuiderkwast) [inline valkey.conf]: Fascinating.

We should have this information also on the RDMA.md documentation page.

**Review by zuiderkwast — APPROVED** [review]
LGTM.

Let's wait for a core team majority decision before  merge since it's an interface change.

Next time, can you avoid force-push and instead just push more commits? It's easier to see what's changed since last review. We squash-merge in the end anyway, normally.

File: valkey.conf
Comment (pizhenwei) [reply valkey.conf]: Sure, glad to do it.

**Comment by pizhenwei** [comment]
> Next time, can you avoid force-push and instead just push more commits? It's easier to see what's changed since last review. We squash-merge in the end anyway, normally.

OK.

**Comment by zuiderkwast** [comment]
@valkey-io/core-team This adds `make BUILD_RDMA=yes` and some configs. Please approve or vote :+1: or :-1:.

The module style configs `rdma.port` and `rdma.bind` are removed. Even if BUILD_RDMA=module, the module is using the new, normal configs. That's why it is a breaking change, but we allow it because RDMA is experimental.

File: README.md
Comment (enjoy-binbin) [inline README.md:213]: so this just a typo?

File: src/config.c
Comment (enjoy-binbin) [inline src/config.c]:
```suggestion
```

File: src/config.c
Comment (enjoy-binbin) [inline src/config.c]:
```suggestion
```

File: src/config.c
Comment (enjoy-binbin) [inline src/config.c]: not that important, just one is three lines and one is one line.
```suggestion
    for (j = 0; j < orig_bindaddr_count; j++) zfree(bindaddr[j]);
```

**Review by enjoy-binbin — APPROVED** [review]
Not a full review, but LGTM.

File: README.md
Comment (pizhenwei) [reply README.md:213]: Previously, this parameter is generated automatically by valkey. valkey server combines module name 'rdma' and module parameter 'port' into 'rdma.port'. Now, we use valkey server parameter 'rdma-port' instead.

Comment (zuiderkwast) [reply README.md:213]: `rdma.port` was the module-defined config, which is removed now because there are normal configs instead.

Before this PR, the RDMA module created module configs in OnLoad. Now, even if RDMA is built as a module, it is using the real configs instead of module-defined config.

The removed module configs is a breaking change, but RDMA is experimental, so it's no problem.

Comment (enjoy-binbin) [reply README.md:213]: so, now if we are using module mode, the old rdma.port is not valid anymore? Yes, it is a breaking change and we should mention it in the release notes (thought it is experimental)

Comment (pizhenwei) [reply README.md:213]: > so, now if we are using module mode, the old rdma.port is not valid anymore? Yes, it is a breaking change and we should mention it in the release notes (thought it is experimental)

Right. `rdma.port` is totally removed.

Comment (zuiderkwast) [reply README.md:213]: > Yes, it is a breaking change and we should mention it in the release notes (thought it is experimental)

It doesn't look good that we have a section of breaking changes in the minor version. We have to write it in a smart way so that it doesn't look bad.

Comment (madolson) [reply README.md:213]: I would just not list it as a breaking change. I would just call it out as a change in the RDMA code, which is still experimental.

Comment (madolson) [reply README.md:213]: I guess alternatively, can we make rdma-port an alias for rdma.port?

File: valkey.conf
Comment (madolson) [inline valkey.conf:303]: We should list this as experimental, like we used to do for cluster mode.

File: valkey.conf
Comment (madolson) [inline valkey.conf]:
```suggestion
# rdma-completion-vector 0
```
Any reason not to be more verbose here? I don't think comp -> completion is a common abbreviation.

I also don't really understand what value you are setting here. If I set 3, am I telling it to only using CPU 3 (or cpus 0 and 1 because that is the bit mask.)

**Review by madolson — COMMENTED** [review]
Directionally I'm inclined, I mostly just had some questions about the configs and documentation.

File: README.md
Comment (pizhenwei) [reply README.md:213]: We have already use `--rdma-bind 192.168.122.100 --rdma-port 6379` instead of `--loadmodule src/valkey-rdma.so port=6379 bind=192.168.122.1` since this change, should we (and other few users in use) really care this dynamic config...

File: valkey.conf
Comment (pizhenwei) [reply valkey.conf]: `comp-vector` is same as [`struct ibv_cq *ibv_create_cq(struct ibv_context *context, int cqe, void *cq_context,  struct ibv_comp_channel *channel,  int comp_vector);`](https://man7.org/linux/man-pages/man3/ibv_create_cq.3.html), this may be friendly to guys  who are familiar with RDMA programming. This is the reason I use `rdma-comp-vector`.

From valkey user side, which one do you prefer? Should I link this man page here?

File: README.md
Comment (madolson) [reply README.md:213]: Yeah, that's true. I'm okay with this as is, I just thought I would bring it up.

File: valkey.conf
Comment (madolson) [reply valkey.conf]:
```
The CQ will use the completion vector comp_vector for signaling completion events
```

Leads me to believe we should use completion-vector, since it seems like the human friendly name. We don't normally abbreviate config names. comp_vector seems like the prototype is just trying to compress it (which might make sense in the context of rdma developers since they will write it a lot).

**Review by madolson — APPROVED** [review]

File: README.md
Comment (zuiderkwast) [reply README.md:213]: If RDMA would not be considered experimental, then we would have had to care about this. It's great that we called it experimental. :smile:

Comment (zuiderkwast) [reply README.md:213]: > I would just not list it as a breaking change. I would just call it out as a change in the RDMA code, which is still experimental.

I removed the breaking-change label again.

**Comment by pizhenwei** [comment]
Hi, @madolson @zuiderkwast 
Append a new commit to fix conflict against 33f42d7fb597ce28040f184ee57ed86d6f6ffbd8. Please take a look at the latest version. Thanks!

**Comment by zuiderkwast** [comment]
@pizhenwei The DCO is failing because some commits are written by some email and signed off with another email. Please fix it. Feel free to squash all commits to one if you want. See DCO job Details: https://github.com/valkey-io/valkey/pull/1209/checks?check_run_id=33460190673

@PingXie @soloestoy @hwware Can we have one more ACK to merge this one? (RDMA is still experimental so it's unclear if it even needs to be a major decision, but it's marked as one now.)

**Comment by hwware** [comment]
> @valkey-io/core-team Please approve.
> 
> This adds `make BUILD_RDMA=yes`, so just like TLS it can be built as a module or as builtin. Just like for TLS, the default is `no`.
> 
> It is a step towards official (non-experimantal) RDMA support.
> 
> I already merged the RDMA fork-safety here: #1244. It makes sure the server doesn't crash if we try to talk to an RDMA client from a forked child process.

Yes, it is the way I recommend. Give user one more option.  Thanks

**Comment by pizhenwei** [comment]
Because of DCO check error, rebase against the latest code and squash into a single commit.

**Pull request merged**
